[WebGL] Use shmem for getBufferSubData over 64 KB

djg opened the pull request:
#### 13571d41220ab567f2f19ebf40c417f4d1b26e06
<pre>
[WebGL] Use shmem for getBufferSubData over 64 KB
<a href="https://bugs.webkit.org/show_bug.cgi?id=275472">https://bugs.webkit.org/show_bug.cgi?id=275472</a>
<a href="https://rdar.apple.com/129831454">rdar://129831454</a>

Reviewed by Kimmo Kinnunen.

RemoteGraphicsContextGLProxy::readPixels is optimized to return data transfers
over 64 KB via shared memory. This change implements the same behavior for
RemoteGraphicsContextGLProxy::getBufferSubData, which is a companion to
readPixels, used to transfer texture data via pixel pack buffers.

Tested by webgl2-large-getbuffersubdata.html which exercises fragmented and
unfragmented transfer for both inline and shared memory transfers.

* LayoutTests/fast/canvas/webgl/webgl2-large-getbuffersubdata-expected.txt: Added.
* LayoutTests/fast/canvas/webgl/webgl2-large-getbuffersubdata.html: Added.
* Source/WebCore/platform/SharedMemory.h:
(WebCore::SharedMemory::subspan const):
(WebCore::SharedMemory::mutableSubspan const):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::getBufferSubDataImpl):
(WebCore::GraphicsContextGLANGLE::getBufferSubData):
(WebCore::GraphicsContextGLANGLE::getBufferSubDataWithStatus):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::getBufferSubDataInline):
(WebKit::RemoteGraphicsContextGL::getBufferSubDataSharedMemory):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(copyBufferSubData):
(getBufferSubData): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::getBufferSubData):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::getBufferSubData): Deleted.
* Tools/Scripts/generate-gpup-webgl:

Canonical link: <a href="https://commits.webkit.org/280157@main">https://commits.webkit.org/280157@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8403ce8aa27a281fe069714da66189569cb6944c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55881 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35205 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8349 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58877 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6313 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58007 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42826 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6509 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/44994 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4352 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57910 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33098 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48176 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26130 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29882 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5504 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4456 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5775 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60466 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5898 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/52425 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48245 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/51925 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8255 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31034 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32118 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33200 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31866 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -276,6 +276,42 @@ void RemoteGraphicsContextGL::simulateEventForTesting(WebCore::GraphicsContextGL
+    }
+    MallocPtr<uint8_t> bufferStore;
+    std::span<uint8_t> bufferData;
+    if (dataSize < getBufferSubDataInlineSizeLimit) {
```
There's a off-by-one where dataSize == getBufferSubDataInlineSizeLimit is not good for anybody

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
```diff
@@ -289,6 +289,48 @@ void RemoteGraphicsContextGLProxy::simulateEventForTesting(SimulatedEventForTest
+        return;
+    }
+inlineCase:
+    auto sendResult = sendSync(Messages::RemoteGraphicsContextGL::GetBufferSubDataInline(target, offset, data.size_bytes()));
```
if the SharedMemory allocation fails, you end up here with size limit bigger than what the receiver will allow. I wonder if this is a problem with texImage2D, too?

It's anyway good idea to iterate and transfer in blocks.

djg replied:
This code is taken from readPixels, so it has the same problem?

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
```diff
@@ -289,6 +289,48 @@ void RemoteGraphicsContextGLProxy::simulateEventForTesting(SimulatedEventForTest
+        auto handle = replyBuffer->createHandle(SharedMemory::Protection::ReadWrite);
+        if (!handle)
+            goto inlineCase;
+        auto sendResult = sendSync(Messages::RemoteGraphicsContextGL::GetBufferSubDataSharedMemory(target, offset, WTFMove(*handle)));
```
Since we're spending time implementing this, we could perhaps limit the shared memory allocation to certain amount and transfer it in blocks. So if client asks for 1gb of data, maybe we still transfer via 100mb blocks to avoid memory limits.

djg replied:
I think the chunking should be added in a follow up patch.

kkinnunen-apple commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
```diff
@@ -895,16 +895,31 @@ void GraphicsContextGLANGLE::bufferSubData(GCGLenum target, GCGLintptr offset, s
 }
 
-void GraphicsContextGLANGLE::getBufferSubData(GCGLenum target, GCGLintptr offset, std::span<uint8_t> data)
+std::span<const uint8_t> GraphicsContextGLANGLE::getBufferSubDataImpl(GCGLenum target, GCGLintptr offset, size_t dataSize, uint8_t* data)
```
dataSize and data come from spans, so span could be used directly.
The return value is gives the impression that it could be a subspan, but it never can be a subspan, as it doesn't make sense. Thus the return value could be a bool.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -276,6 +276,42 @@ void RemoteGraphicsContextGL::simulateEventForTesting(WebCore::GraphicsContextGL
+    }
+    std::span<const uint8_t> validBufferData;
+    if (bufferData.size() == dataSize)
+        validBufferData = m_context->getBufferSubDataWithStatus(target, offset, bufferData);
```
This hunk could be moved inside the ifs above.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -276,6 +276,42 @@ void RemoteGraphicsContextGL::simulateEventForTesting(WebCore::GraphicsContextGL
+    else
+        m_context->addError(GCGLErrorCode::InvalidOperation);
+
+    completionHandler(validBufferData.size());
```
the two cases: inline and shm cannot produce different results o errors.
Currently:
- inline: the memory contains zeros for errors
- shm: returns the buffer size that is valid

So these should somehow be behaving the same

Note: It may be that previous getBufferSubData was copying over the client buffer even when the GPUP would run into an error.

kkinnunen-apple commented on LayoutTests/fast/canvas/webgl/webgl2-large-getbuffersubdata.html:
```diff
@@ -0,0 +1,44 @@
+      const bufferSizeBytes = dim * 4;
+      var buffer = gl.createBuffer();
+      gl.bindBuffer(gl.ARRAY_BUFFER, buffer);
+      gl.bufferData(gl.ARRAY_BUFFER, bufferSizeBytes, gl.STATIC_DRAW);
```
Would it be possible to put some data there?
E.g. sometimes bugs hide in the fact that the verification just verifies that zeros was there.
I typically put i%255...

djg replied:
That's why I filled the receiver with `100`. As can be seen here in the failing test on iOS https://ews-build.s3-us-west-2.amazonaws.com/iOS-17-Simulator-WK2-Tests-EWS/3ce929f9-32909-stress-mode/fast/canvas/webgl/webgl2-large-getbuffersubdata-pretty-diff.html

kkinnunen-apple replied:
These are still different things:
- Verify the get returns zeros
vs
- Verify the get retruns the values that were put in there

kkinnunen-apple commented on Source/WebCore/platform/SharedMemory.h:
```diff
@@ -115,6 +115,20 @@ class SharedMemory : public ThreadSafeRefCounted<SharedMemory> {
     std::span<const uint8_t> span() const { return { static_cast<const uint8_t*>(m_data), m_size }; }
     std::span<uint8_t> mutableSpan() const { return { static_cast<uint8_t*>(m_data), m_size }; }
 
+    std::span<const uint8_t> subspan(size_t offset, size_t count) const
```
This goes against the idea of std::span.
The idea is to obtain the span and then limit it with subspan()

djg replied:
🆗 I just thought that `span(...).subspan()` is ugly

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -276,6 +277,41 @@ void RemoteGraphicsContextGL::simulateEventForTesting(WebCore::GraphicsContextGL
+    bufferStore = MallocPtr<uint8_t>::tryMalloc(dataSize);
+    if (bufferStore)
+        bufferData = { bufferStore.get(), dataSize };
+    if (!m_context->getBufferSubDataWithStatus(target, offset, bufferData))
```
this call should be done only if bufferdata has something?

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -276,6 +277,41 @@ void RemoteGraphicsContextGL::simulateEventForTesting(WebCore::GraphicsContextGL
+    MallocPtr<uint8_t> bufferStore;
+    std::span<uint8_t> bufferData;
+    if (dataSize >= getBufferSubDataInlineSizeLimit)
+        RELEASE_LOG(WebGL, "getBufferSubData requested %zu bytes of inline data", dataSize);
```
This can be now hard requirement, as you implemented the chunking?

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
```diff
@@ -289,6 +289,63 @@ void RemoteGraphicsContextGLProxy::simulateEventForTesting(SimulatedEventForTest
+    static constexpr size_t getBufferSubDataSharedMemorySizeLimit = 100 * MB;
+
+    auto copyToData = [&](ptrdiff_t offset, std::span<const uint8_t> replyData) {
+        memcpy(data.data() + offset, replyData.data(), replyData.size_bytes());
```
idea of std::span is that you don't anymore carry offsets and lenghts around. instead you have your span, and you cut out the unneeded parts with subspan.
Also std::copy might already be able to do the copy

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
```diff
@@ -289,6 +289,63 @@ void RemoteGraphicsContextGLProxy::simulateEventForTesting(SimulatedEventForTest
+
+    ptrdiff_t transferOffset = 0;
+    size_t transferRemaining = data.size();
+    size_t transferSize = 0;
```
transferSize can and should be  a variable local to the scope it's used in.

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
```diff
@@ -289,6 +289,63 @@ void RemoteGraphicsContextGLProxy::simulateEventForTesting(SimulatedEventForTest
+        memcpy(data.data() + offset, replyData.data(), replyData.size_bytes());
+    };
+
+    ptrdiff_t transferOffset = 0;
```
should be replaced with span that you adjust

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
```diff
@@ -289,6 +289,63 @@ void RemoteGraphicsContextGLProxy::simulateEventForTesting(SimulatedEventForTest
+            if (!valid)
+                return;
+            copyToData(transferOffset, replyBuffer->subspan(0, transferSize));
+            transferOffset += transferSize;
```
this would be something like `data = data.subspan(transferSize)`

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
```diff
@@ -289,6 +289,63 @@ void RemoteGraphicsContextGLProxy::simulateEventForTesting(SimulatedEventForTest
+            markContextLost();
+            return;
+        }
+        auto [inlineBuffer] = sendResult.takeReply();
```
here you need a similar early return as with !valid case above.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -276,6 +277,41 @@ void RemoteGraphicsContextGL::simulateEventForTesting(WebCore::GraphicsContextGL
+
+    handle.setOwnershipOfMemory(m_sharedResourceCache->resourceOwner(), WebKit::MemoryLedger::Default);
+    if (auto buffer = SharedMemory::map(WTFMove(handle), SharedMemory::Protection::ReadWrite))
+        validBufferData = m_context->getBufferSubDataWithStatus(target, offset, buffer->mutableSubspan(0, dataSize));
```
The malicious actor might send too high dataSize, so you need to validate that the dataSize fits into the buffer.
This is, when you remove the mutableSubspan from SharedMemory.
That checking is MESSAGE_CHECK

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -44,6 +45,8 @@
 #include "RemoteVideoFrameObjectHeap.h"
 #endif
 
+#define MESSAGE_CHECK(assertion) MESSAGE_CHECK_BASE(assertion, (m_gpuConnectionToWebProcess.get()->protectedConnection().ptr()))
```
Ah, we didn't have a instance of message check here.
This is no good, as m_gpuConnectionToWebProcess is not thread-safe.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -276,6 +279,48 @@ void RemoteGraphicsContextGL::simulateEventForTesting(WebCore::GraphicsContextGL
+    assertIsCurrent(workQueue());
+    static constexpr size_t getBufferSubDataInlineSizeLimit = 64 * KB; // NOTE: when changing, change the value in RemoteGraphicsContextGLProxy too.
+
+    MESSAGE_CHECK(dataSize <= getBufferSubDataInlineSizeLimit);
```
maybe just have a early return if dataSize > getBufferSubDataInlineSizeLimit

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -276,6 +279,48 @@ void RemoteGraphicsContextGL::simulateEventForTesting(WebCore::GraphicsContextGL
+void RemoteGraphicsContextGL::getBufferSubDataSharedMemory(uint32_t target, uint64_t offset, size_t dataSize, WebCore::SharedMemory::Handle handle, CompletionHandler<void(bool)>&& completionHandler)
+{
+    assertIsCurrent(workQueue());
+    bool validBufferData  = false;
```
double space here

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -276,6 +279,48 @@ void RemoteGraphicsContextGL::simulateEventForTesting(WebCore::GraphicsContextGL
+    MESSAGE_CHECK(dataSize <= readPixelsSharedMemorySizeLimit);
+
+    handle.setOwnershipOfMemory(m_sharedResourceCache->resourceOwner(), WebKit::MemoryLedger::Default);
+    if (auto buffer = SharedMemory::map(WTFMove(handle), SharedMemory::Protection::ReadWrite))
```
this is still lacking the size check.
E.g. if caller sends 4kb memory and 50mbs dataSize, it's not going well.

djg committed:
[WebGL] Use shmem for getBufferSubData over 64 KB
https://bugs.webkit.org/show_bug.cgi?id=275472
rdar://129831454

Reviewed by Kimmo Kinnunen.

RemoteGraphicsContextGLProxy::readPixels is optimized to return data transfers
over 64 KB via shared memory. This change implements the same behavior for
RemoteGraphicsContextGLProxy::getBufferSubData, which is a companion to
readPixels, used to transfer texture data via pixel pack buffers.

Tested by webgl2-large-getbuffersubdata.html which exercises fragmented and
unfragmented transfer for both inline and shared memory transfers.

* LayoutTests/fast/canvas/webgl/webgl2-large-getbuffersubdata-expected.txt: Added.
* LayoutTests/fast/canvas/webgl/webgl2-large-getbuffersubdata.html: Added.
* Source/WebCore/platform/SharedMemory.h:
(WebCore::SharedMemory::subspan const):
(WebCore::SharedMemory::mutableSubspan const):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::getBufferSubDataImpl):
(WebCore::GraphicsContextGLANGLE::getBufferSubData):
(WebCore::GraphicsContextGLANGLE::getBufferSubDataWithStatus):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::getBufferSubDataInline):
(WebKit::RemoteGraphicsContextGL::getBufferSubDataSharedMemory):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(copyBufferSubData):
(getBufferSubData): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::getBufferSubData):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::getBufferSubData): Deleted.
* Tools/Scripts/generate-gpup-webgl:

Canonical link: https://commits.webkit.org/280157@main
